### PR TITLE
support filter in HomePageVisitedByType

### PR DESCRIPTION
**Pull request by shmaram**
## Hey, I just made a Pull Request!

Added filter support for HomePageVisitedByType in order to enable filtering entites from the list

Fixes: [21243](https://github.com/backstage/backstage/issues/21243)

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/home/src/homePageComponents/VisitedByType/Content.tsx
Comment (djamaile) [inline plugins/home/src/homePageComponents/VisitedByType/Content.tsx:74]: We can this a bit shorter and easier to read:

```typescript
      return await visitsApi
      .list({
        limit: numVisitsTotal ?? 8,
        orderBy: [{ field: 'timestamp', direction: 'desc' }],
        ...(filterBy && { filterBy }),
      })
      .then(setVisits);
```
      
Same for below

**Review by djamaile — CHANGES_REQUESTED** [review]
You will also need to run the api report command to pass the pipeline

`yarn build:api-reports plugins/home/`

**Comment by shmaram** [comment]
Thanks @djamaile - Fixed all comments 😄

**Review by shmaram — COMMENTED** [review]
fixed :)

File: .changeset/wise-flies-laugh.md
Comment (emmaindal) [inline .changeset/wise-flies-laugh.md:2]: I thiiink as `filterBy` is optional, this can be a patch. I'll just double check with @backstage/maintainers

**Review by emmaindal — COMMENTED** [review]
Awesome, just double checking the changeset with @backstage/maintainers before we get this in!

**Comment by tikabom** [comment]
@shmaram Can I help with the changes required to complete this contributions?

**Comment by shmaram** [comment]
@tikabom - I will work on it in the next few days

**Comment by shmaram** [comment]
@tikabom - Done, any idea why the `Docs quality` failed?

File: plugins/home/README.md
Comment (awanlin) [inline plugins/home/README.md:338]: Vale isn't happy with `entityRef`, if you put backticks (`) around it that is one way to fix it

**Review by tikabom — APPROVED** [review]
Amazing work! Thanks for receiving all the feedback and taking the effort to improve the features of this plugin.

File: plugins/home/src/homePageComponents/VisitedByType/Content.test.tsx
Comment (djamaile) [reply plugins/home/src/homePageComponents/VisitedByType/Content.test.tsx:149]: It would be nice to cover the following four scenarios:

```
when kind === ‘recent’ and home.recentVisits.filterBy is available
when kind === ‘recent’ and home.recentVisits.filterBy is not available
when kind === ‘top’ and home.topVisits.filterBy is available
when kind === ‘top’ and home.topVisits.filterBy is not available
```

Comment (shmaram) [reply plugins/home/src/homePageComponents/VisitedByType/Content.test.tsx:158]: So I was looking at the test again, we mock the visits API so regardless of the filter I will add in the config the response will be the same as the mock. So I updated the test to check the parameters that `list` is receiving. 
WDYT?

Comment (shmaram) [reply plugins/home/src/homePageComponents/VisitedByType/Content.test.tsx:158]: @djamaile , any chance you can review my changes?

Comment (tikabom) [reply plugins/home/src/homePageComponents/VisitedByType/Content.test.tsx:158]: unit tests look good to me.

File: plugins/home/src/api/config.ts
Comment (tikabom) [inline plugins/home/src/api/config.ts:45]: I have a concern here.
I would expect that throwing this error here breaks the UI?

We want to capture the invalid config during the build and report that back to the user who is configuring the filters. 

And here instead of throwing an error, we probably just want to default to no filtering.

Comment (shmaram) [reply plugins/home/src/api/config.ts:45]: I can change the catch implementation to use "no filtering". 
I'm not sure how to capture the invalid config during the build, is there an example?

Comment (shmaram) [reply plugins/home/src/api/config.ts:45]: This will not break the UI because this code is running inside `readFilterByConfigs` and incase an error was thrown we will return undefined there 

`export function readFilterByConfigs(
  configs: Config[],
): VisitsApiQueryParams['filterBy'] | undefined {
  try {
    return configs.map(readFilterByConfig);
  }
  catch {
    return undefined;
  }
}`

then, in `context.tsx` we get the value and use it there.

> We want to capture the invalid config during the build and report that back to the user who is configuring the filters.

I'm not sure I get this one, currently, I saw an error in the log but the UI doesn't break

@tikabom - WDYT?

Comment (emmaindal) [reply plugins/home/src/api/config.ts:45]: @shmaram will let @tikabom answer your questions but just noting that we don't have tests for these helper functions `readFilterByConfig` and `readFilterByConfigs`. Can we add that as part of this PR? that could also test the case of invalid config by the way!

Comment (shmaram) [reply plugins/home/src/api/config.ts:45]: @emmaindal - Added your suggested tests. 
Will wait for @tikabom to review as well.

BTW - does anyone know why the "E2E Linux" check failed?

Comment (emmaindal) [reply plugins/home/src/api/config.ts:45]: Hello @shmaram sorry for late replies here just coming back after some time off. Some test fails due to prettier errors in 

```
plugins/home/src/api/config.test.ts
plugins/home/src/api/config.ts
```

can you run prettier against these files? `yarn prettier --write plugins/home/src/api`

Comment (tikabom) [reply plugins/home/src/api/config.ts:45]: @shmaram my bad! I didn't realise that `readFilterByConfig` was being called by `readFilterByConfigs` which is then called in `Content.tsx`! Could we rename these functions to `readFilterConfig` and `createFilterByQueryParamFromConfig`?

Another question here: Do we want to ignore all filter configs even if only a subset of them are bad? For example if there were 3 different filter configs and if only one of them were bad, throwing an error would mean that `filterBy` would always be `undefined`. What do you think @emmaindal ? Should we still be accepting the correctly configured filters?

Also it seems like bad app-configs aren't automatically captured as part of `yarn build` and need to be captured explicitly by running `@backstage/cli config:check`. @shmaram Could you add this to the README as a note? Thank you!

Comment (tikabom) [reply plugins/home/src/api/config.ts:45]: I agree with you.. @shmaram let us know if you have any specific use cases in mind why it makes sense to ignore all filters.

Comment (shmaram) [reply plugins/home/src/api/config.ts:45]: Hi @emmaindal @tikabom 
Thanks for the comments! 
The reason that I thought we should ignore all filters is to avoid misunderstanding for the developer, let's say they add 10 filters and 1 was bad, it would be very difficult to know that 9 are good but 1 didn't pass, but if no filters were set than it would be easy to see. 
Your suggestion works well also, I can go with both, so I will let you choose :)

Comment (tikabom) [reply plugins/home/src/api/config.ts:45]: Hi @shmaram, let's apply filters that are set correctly. Since each filter config is processed separately anyway, you can throw the Error with the Config object currently being processed.

That + README update, and you're good to 🚢

Comment (shmaram) [reply plugins/home/src/api/config.ts:45]: Done 😄

File: plugins/home/src/api/config.ts
Comment (tikabom) [inline plugins/home/src/api/config.ts:45]: If we want to ignore the badly configured filter shouldn't this be returning undefined? If an error is thrown here, you will end up in line 62 which is just no filters? Maybe I phrased myself poorly in my previous comment. I was saying we should accept filters that are configured correctly and ignore the ones that are not.

File: plugins/home/README.md
Comment (tikabom) [inline plugins/home/README.md:356]:
```suggestion
`filterBy` configs that are not defined in the above format will be ignored. 
```

File: plugins/home/src/api/config.ts
Comment (freben) [inline plugins/home/src/api/config.ts:42]: The schema says string or number, but this forces string.

File: plugins/home/src/api/config.ts
Comment (freben) [inline plugins/home/src/api/config.ts:62]: Is it ok to give up entirely silently? This'll also swallow accidental typos in the config etc.

File: plugins/home/src/api/config.ts
Comment (freben) [inline plugins/home/src/api/config.ts:41]: Just noting that this doesn't do any runtime validation that the value actually is one of those.

**Review by freben — COMMENTED** [review]
Just saw a few small things as I passed by this, @djamaile for final review

File: plugins/home/src/api/config.ts
Comment (shmaram) [reply plugins/home/src/api/config.ts:67]: Yes, this is what we agreed on (also mentioned in the readme)

https://github.com/backstage/backstage/pull/21470#discussion_r1466578836

File: plugins/home/src/api/config.ts
Comment (shmaram) [reply plugins/home/src/api/config.ts:43]: Created Operators type and added type guard

**Review by freben — APPROVED** [review]
Approved from my side

File: plugins/home/src/api/config.ts
Comment (tikabom) [reply plugins/home/src/api/config.ts]: great catch @freben ! thank you

**Review by tikabom — APPROVED** [review]
Grande!! 🚢

**Review by djamaile — APPROVED** [review]

**Comment by heungju-lee** [comment]
i am testing this filter feature, it doesn't work at all except with the below...

`home:
  recentVisits:
    filterBy:
      - field: entityRef
        operator: "contains"
        value: "component"
  topVisits:
    filterBy:
      - field: entityRef
        operator: "contains"
        value: "component"
`

the other operator was not working with entityRef and any value.
and i want not only 'component' but also other value such as 'API', 'Resource'...
for now, i could use only one value to filter.
I don't know if this was intentional.

**Pull request merged**
